test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java

```diff
@@ -0,0 +1,271 @@
+                }
+                result = artifactPropertiesDirectory.toPath();
+            }
+        }
+        if (result == null) {
+            throw new IllegalStateException(
+                    "Unable to locate the artifact metadata file created that must be created by Quarkus in order to run tests annotated with '@QuarkusIntegrationTest'.");
+        }
+        if (!Files.isDirectory(result)) {
+            throw new IllegalStateException(
+                    "The determined Quarkus build output '" + result.toAbsolutePath().toString() + "' is not a directory");
+        }
+        return result;
+    }
+
+    private static File determineBuildOutputDirectory(final URL url) {
+        if (url == null) {
+            return null;
+        }
+        if (url.getProtocol().equals("file") && url.getPath().endsWith("test-classes/")) {
+            //we have the maven test classes dir
+            File testClasses = new File(url.getPath());
+            return testClasses.getParentFile();
+        } else if (url.getProtocol().equals("file") && url.getPath().endsWith("test/")) {
+            //we have the gradle test classes dir, build/classes/java/test
+            File testClasses = new File(url.getPath());
+            return testClasses.getParentFile().getParentFile().getParentFile();
+        } else if (url.getProtocol().equals("file") && url.getPath().contains("/target/surefire/")) {
+            //this will make mvn failsafe:integration-test work
+            String path = url.getPath();
+            int index = path.lastIndexOf("/target/");
+            return new File(path.substring(0, index) + "/target/");
+        }
+        return null;
+    }
+
+    private IntegrationTestExtensionState doProcessStart(Properties quarkusArtifactProperties,
+            Class<? extends QuarkusTestProfile> profile, ExtensionContext context)
+            throws Throwable {
+        quarkusTestProfile = profile;
+        TestResourceManager testResourceManager = null;
+        try {
+            Class<?> requiredTestClass = context.getRequiredTestClass();
+
+            Map<String, String> sysPropRestore = getSysPropsToRestore();
+            Map<String, String> additionalProperties = determineAdditionalProperties(profile,
+                    sysPropRestore);
+
+            testResourceManager = new TestResourceManager(requiredTestClass);
+            testResourceManager.init();
+            additionalProperties.putAll(testResourceManager.start());
+
+            String artifactType = quarkusArtifactProperties.getProperty("type");
+            if (artifactType == null) {
+                throw new IllegalStateException("Unable to determine the type of artifact created by the Quarkus build");
+            }
+            ArtifactLauncher launcher;
+            switch (artifactType) {
+                case "native": {
```

Comment:
This is ok for now, but I think long term we want the extension that contributes the artifact to also provide the launcher somehow to make this extensible.

Reply:
Definitely. I was thinking of having a Service loader.
Once that need arises, I'll change the code ASAP